Implemented support for Python 3

KonstantinShemyak opened the pull request:
This pull request makes the project compatible with Python 3.
Closes #53 .

All tests [pass on Python 2.7 and Python 3.6](https://travis-ci.org/KonstantinShemyak/python-javatools/builds/436566138).

No logic is changed and the attempt has been made to isolate the language differences. Some checks might be refactored further (grep for `isinstance`).

I aimed to abide to these principles:

- test for functionality, not for version (`if isinstance()`, not `if PYTHON_VERSION == X`)
- separate the text/binary data as early as possible on input, convert it back as late as possible on output

When(/if) this PR gets accepted, it would be possible to celebrate:

- update `ChangeLog`, `README.md`
- declare Python3 support in the trove classifier.

Additionally it would be nice to have in-project testing framework for both Py2 and Py3.

obriencj commented on javatools/__init__.py:
```diff
@@ -2093,10 +2098,7 @@ def _pretty_const_type_val(typecode, val):
-            val = repr(val)[2:-1]  # trim off the surrounding u"" (HACK)
-        else:
-            val = repr(val)[1:-1]  # trim off the surrounding "" (HACK)
+        val = val.encode('utf-8')
```
the point of spitting out a repr was to escape certain characters, not just to encode.

KonstantinShemyak replied:
* Dependency on `future` and `past`:

In fact `past` can be avoided. (By replacing all `xrange` with `range`; this does not really have an impact on Py2, as our input data is small - definitely not at size where difference between generator and list could be noticed. On Py3, `range` is the only one left (I like [this writing](http://treyhunner.com/2018/02/python-3-s-range-better-than-python-2-s-xrange/) on the subject)).

I have amended the commit where I had imported `past`, rebased and force-pushed the branch. 
(BTW, **every** commit of this rather large push request passes **all** tests in Py2.)

KonstantinShemyak replied:
_> the point of spitting out a repr was to escape certain characters, not just to encode._

Reverted back to what you had, but in a form which compiles under Py3. Commit amended, branch force-pushed. (Please tell if you find force-pushing of PRs' branches inconvenient.)

[2.7 and 3.6 tests pass](https://travis-ci.org/KonstantinShemyak/python-javatools/builds/436706619), although these lines were obviously not covered.

In fact our Unicode support is not tested. I had earlier https://github.com/KonstantinShemyak/python-javatools/issues/34 and just created https://github.com/KonstantinShemyak/python-javatools/issues/42 and https://github.com/KonstantinShemyak/python-javatools/issues/43. But I want to get Py3 otherwise supported before checking these issues, so that they would be handled in both language versions.

KonstantinShemyak replied:
For cleaner history, I have squashed the latter change into the commit with message "Python3-compatible syntax", so there is no incorrect code in any commit in the chain.

obriencj commented:
yup, I'm familiar with some of the simple py2 and py3 compatability tweaks
eg. https://github.com/obriencj/python-mapbind/blob/master/mapbind/__init__.py#L33

I'd much prefer those as needed vs. reliance on additional packages. If I could go back in time I'd suggest that you do this in stages, first with replacing print with the print function, etc, just to make it easier to step through the changes. I think (hope) that the buffer to string changes will be safe, still reviewing other aspects though.

KonstantinShemyak commented:
> I'd suggest that you do this in stages, first with replacing print with the print function, etc, just to make it easier to step through the changes.

The pull request is done exactly in this way:

- Every commit leaves the repository in fully consistent state (all Py2 tests pass)
- Every commit has clear purpose, which is stated in its message.

They can be merged one-by-one if wanted, but naturally only the last one makes also all tests pass on Py3.

obriencj commented:
Hrm, was trying to review individual commits, looks like github considered that as a review of the whole PR. And only the whole PR can be merged as a unit.

obriencj commented:
I fixed some issues with the print function changes, merged to master and rebased this on top.

obriencj commented:
The manifest max line changes were unnecessary and do nothing but add repeated math calls to the parser. I agree that magic numbers aren't the cleanest, but simply commenting as to why the number was there would have been better.

KonstantinShemyak commented on javatools/__init__.py:
```diff
@@ -1888,7 +1890,7 @@ def __eq__(self, other):
 
         # for each of the key/val pairs, check equality
-        for key, lval in self.items():
+        for key, lval in items(self):
```
Whoops. That's true, this change is not needed.
I have corrected its commit, amended, and (again) rebased the `py3` branch on top of it.

I have not yet rebased `py3` on the recent `master` update, as the latter fails tests (because of minor mistype).

KonstantinShemyak commented:
Ups. I understood my problem with missing `from __future__ import print_function`. The code **worked** in Python 2 without this import, because it treated the argument as a tuple!
Corrected, rebased the branch (on the way, also updated some commit messages).

obriencj commented:
I've spent some time thinking about this, and I think the best way to proceed here would be

1. from master, start using print_function (done), six.renames, and six.add_metaclass. The six module has an extremely wide distribution and is packaged up all over the place. This can be a pull request on its own, and can include renames of assertEqual plus using memoryview on py3 where buffer raises a NameError. There are some changes you made that I would do very differently, like how you switched to using range everywhere, and how you're using future and past. Introduce dependency changes here and bump to 1.5.0. This PR should be almost nothing but potential compatability with names.

2. material changes to API, helper functions that avoid using incremental reads. I saw a few of these in manifest and class parsing for example, places where the normal path went from reading from a stream to reading all into memory and then traversing that. I don't necessarily have an issue with this change, but it seems like a separate issue.

3. leftover changes needed to make it run under python3. change the setup.py to reflect the versions that are supported. This would then mark a 1.5.0 release

KonstantinShemyak commented:
I'll close this PR now as it does not apply on the current master, and as I realized that some reductions of the API (not accepting streams) are not strictly necessary. _(Side note: my personal taste is against such "promiscuous" interfaces in cases when they have to be handled by enumerating `if isinstance()...` cases. But, well, that's just a personal taste of a contributor :-) )_

I have identified the following differences in text/binary I/O:

* `ZipFile` returns **filenames list** from the zip (jar) in `str` or in `unicode` in Py2, depending on the flag in the zip file. In Py3, it always returns `str`.
* `ZipFile` returns **data** from zip archives as **binary**, which is `str` in Py2 (and currently handled as such) but `bytes` in Py3. That data itself represents different things:
   * Java **method names**, **class names** and other identifiers are UTF-8-encoded strings.
   * Jar **manifest** and **signature file** are UTF-8 encoded texts.
   * Other Java **class data** can be treated as "just binary data" as far as we are not processing it in any particular way.
* `b64encode()` returns `bytes` on Py3 and `str` on Py2. The content is in fact ASCII string.

I might re-implement these differences one-by-one some time soon.

obriencj commented:
six.text_type and six.binary_type might be the answer to some of this, though there might still be some annoying conversion steps

obriencj commented:
All in all, this is definitely looking very possible, and I'll happily support the more intense changes you were looking to make on top of current master. If there's additional prepwork needed on top of master, let me know.
